TS Definition to Hapi v17

rafaelsouzaf opened the pull request:
First commits with new TypeScript definitions for hapi v17.

- [X ] Use a meaningful title for the pull request. Include the name of the package modified.
- [ ] Test the change in your own code. (Compile and run.)
- [ X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [X ] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [ ] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [X ] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/hapijs/hapi/blob/master/API.md
- [X ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

If removing a declaration:
- [ ] If a package was never on DefinitelyTyped, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

BorntraegerMarc commented on types/hapi/v17/hapi-tests.ts:
```diff
@@ -0,0 +1,7 @@
+import Hapi = require("hapi");
```
I would import it here with `import { Server } from 'hapi'` Then instanciate it with:
```
import { Server } from 'hapi';
const server = new Server({});
```

rafaelsouzaf replied:
Oh yes. I didn't work with this file yet. It's your original code from "hapi-17.0-marc" branch. I can remove this file from the PR if you want. What do you think?

BorntraegerMarc commented on types/hapi/v17/index.d.ts:
```diff
@@ -0,0 +1,15 @@
+import Events = require("events");
+import url = require("url");
+
+export * from './definitions/server/server'
```
Aren't the other exports missing?

rafaelsouzaf replied:
Of course, is almost the same here. I didn't work with this file yet. I only added your profile in "Definitions by" I guess.

rafaelsouzaf commented:
@BorntraegerMarc 

> Did you decide now to do TS 2.6 with Hapi v16 first and then move to Hapi v17?

Oh no. Never. I am working with v17 and TS 2.6 only.

> Because in this PR there are also a lot of Hapi v16 changes.

It's because I am following the recommendation of @AJamesPhillips in the issue:
(https://github.com/DefinitelyTyped/DefinitelyTyped/issues/20670#issuecomment-337924152)

To start working with the v17, first...

> "... we'd make a v16 folder first and copy the current index.d.ts and tests to that, then update the root level index.d.ts, tests and tsconfig to work with typescript 2.6 and hapi v17."

So I created the v16 folder, moved v16 files to inside and changed the paths to keep working fine :-)
I think we talked about that before. Maybe I am not understanding very well the recommendation.

> I would maybe suggest that we jump directly to v17 and leave v16 be (because it seems to work well for now...) Sorry for changing my opinion again :)

No problem. I liked :)

> But definitely the v16 changes need to be deleted from this PR. If we want to do them it needs to be in a separate PR

Ok no problem. Just to confirm: So, I need to keep in this PR only the v17 folder, right?

BorntraegerMarc commented:
@rafaelsouzaf 

>  So, I need to keep in this PR only the v17 folder, right?
Yes 😄

rafaelsouzaf commented:
Thanks @BorntraegerMarc I appreciate if you can review the branch now. Please let me know if you have any question.

BorntraegerMarc commented on types/hapi/v17/definitions/plugin/plugin.d.ts:
```diff
@@ -0,0 +1,50 @@
+    [pluginName: string]: any;
+}
+
+// TODO Similar to v16, I am using that interface without any definition.
```
is this really a todo?

rafaelsouzaf replied:
Fixed. Please see the new PR (https://github.com/rafaelsouzaf/DefinitelyTyped/pull/2/)
Sorry. I should not have closed this one.

BorntraegerMarc commented on types/hapi/v17/definitions/request/request.d.ts:
```diff
@@ -0,0 +1,311 @@
+
+/**
+ * TODO both objects ReplyValue and _ReplyValue I found in the v16 TS definition, but I don't found it in the documentation. Need review.
+ * TODO I think it's not very elegant solution
```
You could define the type as `any` in worst case....
Should we post it in https://github.com/hapijs/hapi/issues/3669 ?
And did you also not find it in the tests?

rafaelsouzaf replied:
Yes, maybe we can post there. But I would like to advance a little more. Some parts of the code and API are more "clear" when I review it. I will keep the TODO for now.

BorntraegerMarc commented on types/hapi/v17/definitions/request/request.d.ts:
```diff
@@ -0,0 +1,311 @@
+     * Note that the route's authentication mode and strategies are ignored. The only match is made between the request.auth.credentials scope and entity information and the route access configuration.
+     * If the route uses dynamic scopes, the scopes are constructed against the request.query, request.params, request.payload, and request.auth.credentials which may or may not match between the route and the request's route. If this method is called using a request that has not been authenticated (yet or not at all), it will return false if the route requires any authentication.
+     */
+    // TODO how to implement it? It is a function inside route.auth! The v16 doesn't have this option.
```
Could you define it possibly as `() => Object`?

rafaelsouzaf replied:
Can you give me an example please?

BorntraegerMarc commented on types/hapi/v17/definitions/route/route-options-payload.d.ts:
```diff
@@ -0,0 +1,111 @@
+export enum PayloadMultipartOutputEnum {
+    // TODO not ready. The v16 definition have similar doubts: "// TODO type this?"
+    // https://github.com/hapijs/hapi/blob/master/API.md#route.options.payload.multipart
+    Data = 'data',
```
Don't you think it should look a  bit more like:
```
export enum PayloadMultipartOutputEnum {
  output {
    Data = 'data',
    Stream = 'stream',
    File = 'file',
    Annotated : {
      headers: '',
      filename: '',
      payload: ''
    }
  }
}
```

however the `Annotated` options can be whatever the function caller chooses. So maybe an interface is better fit....

rafaelsouzaf replied:
I don't know if we can use Enum with this format. Are you sure?

With Java I know very well when we should use Enum or not. With TS I am a little bit confused yet because I think some times we can use Interface or Type with the same objective. I am using a lot of Enums in the definitions, but if you like we can review it and maybe refactoring everything.

SimonSchick commented on types/hapi/v17/definitions/server/server-connection-options.d.ts:
```diff
@@ -0,0 +1,54 @@
+import {RouteOptions} from "../route/route-options";
+
+/** Any connections configuration server defaults can be included to override and customize the individual connection. */
+export interface IServerConnectionOptions extends IConnectionConfigurationServerDefaults {
```
What's with the inconsistent interface names?

rafaelsouzaf replied:
@SimonSchick The inconsistent names between for example RouteOptions (starting without "I") and IServerConnectionOptions (starting with "I") right? Yes, we are working on that :-)

SimonSchick replied:
Hah that's fine, I read this diff late at night and thought this already went into DT main repo 😛

rafaelsouzaf replied:
@SimonSchick when you want, please check the PR (https://github.com/rafaelsouzaf/DefinitelyTyped/pull/2/)

gitawego commented:
it seems the param ``value`` for method ``response`` in response-toolkit.d.ts should be ``any`` or at least ``string | object`` imo, depends on [this test](https://github.com/hapijs/hapi/blob/4144f9397cf639e22379699cdfc967d2c84e8509/test/response.js#L220)

rafaelsouzaf commented:
[Fixed](https://github.com/rafaelsouzaf/DefinitelyTyped/pull/1/commits/c739273012528ec666398c5366ecdc55861794e9). Thanks @gitawego 

In the same file we have the method `redirect()` with **no-optional (mandatory)** param in the [JSDoc](https://github.com/hapijs/hapi/blob/master/API.md#-hredirecturi), but optional in the [tests](https://github.com/hapijs/hapi/blob/4144f9397cf639e22379699cdfc967d2c84e8509/test/response.js#L846). Do you know what I should consider?

Maybe the best approach to test everything is migrate the official hapijs tests to Typescript.


**EDITED**: The official [hapijs tests](https://github.com/hapijs/hapi/tree/4144f9397cf639e22379699cdfc967d2c84e8509/test) are huge! I can't migrate to TS everything alone. It will take very long time to do this job and maybe it's not necessary. If anybody can help I will appreciate :) For now I will continue with the basic tests using examples from the doc API and based in v16 version.

ctaggart commented:
I'm currently thinking that it is better to use union types as recommended
https://www.typescriptlang.org/docs/handbook/declaration-files/do-s-and-don-ts.html

Instead of:
``` ts
     response(value: string): ResponseObject;
     response(value: object): ResponseObject;
```

do this:
``` ts
     response(value: string|object): ResponseObject;
```

rafaelsouzaf commented:
@ctaggart I completely understand and I have read this link and recommendation before.

But, on the other hand, there is another recommendation from [definitelytyped.org](http://definitelytyped.org/guides/best-practices.html). Please see _Function Overloading_ sub-section.

> For functions / constructors use function overloading e.g

```
declare class Foo {
    constructor(foo: number);
    constructor(foo: string);
}
```

So I am confused:-)
What is the best option? I am not sure.

PS: @ctaggart If you can help me reviewing this [PR](https://github.com/rafaelsouzaf/DefinitelyTyped/pull/4) I will appreciate :-)

ctaggart commented:
My plan for review of this API was to try it out from Fable by running it through [ts2fable](https://www.npmjs.com/package/ts2fable) which I've been writing. http://blog.ctaggart.com/search/label/ts2fable

I like `response(value: string|object): ResponseObject;` because that is what is going to work. :-)

ctaggart commented:
After sleeping on it, I'll modify ts2fable to support what you have already with:
``` ts
     response(value: string): ResponseObject;
     response(value: object): ResponseObject;
```

rafaelsouzaf commented:
LOL great @ctaggart !
Don't kill me if soon we change to `response(value: string | object): ResponseObject` :-)

ctaggart commented:
> Don't kill me if soon we change to response(value: string | object): ResponseObject :-)

👍 I'll make it work either way.

gitawego commented:
@rafaelsouzaf  for ``redirect()`` part, I think we should be based on the test, if their test passes, it means the param ``uri`` should be optional. 

I will try to check as much as I can their test files, and see if there should be any improvement.  but I think your definition is mature enough to send a PR to upstream, once it merged, ppl could contribute, as you said, there are too many tests, you can't do it alone ^^

gitawego commented:
@rafaelsouzaf  for ``response`` part, depends on [official best practice](https://www.typescriptlang.org/docs/handbook/declaration-files/do-s-and-don-ts.html), in ``Use Union Types`` section, it says: ``Don’t write overloads that differ by type in only one argument position``, so I think it's better with ``response(value: string | object): ResponseObject``

rafaelsouzaf commented:
@gitawego commited [redirect()](https://github.com/rafaelsouzaf/DefinitelyTyped/pull/1/commits/81e58330cca9951cfc9eb9781a6295957af385fb) without params 👍 

Thanks for the tests. I made some, _maybe delicate_, changes today ([1](https://github.com/rafaelsouzaf/DefinitelyTyped/pull/1/commits/e189cf6b5e09901f6c068fc78960e37fa129ec77), [2](https://github.com/rafaelsouzaf/DefinitelyTyped/pull/1/commits/9fb4ab18b394a9aec15afee86d80588dbba834f1)). If you detect any problem please let me know.

I would like to hear the opinion of my mentor @BorntraegerMarc  :-)

rafaelsouzaf commented:
Ok. A lot of tests were built based on v17 API code samples. And some v16 tests have been adapted to the new version.
It's pending the `response(value: string | object)` change.

BorntraegerMarc commented on types/hapi/v17/test/request/event-types.ts:
```diff
@@ -0,0 +1,76 @@
+    }
+};
+
+const onRequest: Lifecycle.Method = function (request: Request, h: ResponseToolkit) {
```
Why is this not an arrow function? `(request: Request, h: ResponseToolkit) => {`

BorntraegerMarc replied:
In general there are quite a bit of "normal" functions. In my opinion the tests should follow one standard. Ether `function` or `() =>`    

I  personally like the newer arrow function standard better 😄

rafaelsouzaf replied:
Commit 1a5f4f5

BorntraegerMarc commented on types/hapi/v17/test/response/response-events.ts:
```diff
@@ -0,0 +1,31 @@
+
+    const response:ResponseObject = request.response;
+    // if (response.isBoom) {
+    //     return null;
```
Comments probably can be deleted?

rafaelsouzaf replied:
Commit 6a11db6

BorntraegerMarc commented on types/hapi/v17/test/route/config.ts:
```diff
@@ -0,0 +1,54 @@
+};
+
+server.route({method: 'GET', path: '/user', options: user });
+//server.route({ method: 'GET', path: '/public/{path*}', options: { cache: { privacy: 'public', expiresIn: 24 * 60 * 60 * 1000 } }, handler: { directory: { path: __dirname, listing: false, index: false } } });
```
Can this comment be deleted?

rafaelsouzaf replied:
Commit 6a11db6

BorntraegerMarc commented on types/hapi/v17/test/server/server-cache-provision.ts:
```diff
@@ -0,0 +1,21 @@
+cache.set('norway', 'oslo', 10 * 1000, null);
+const value = cache.get('norway', null);
+
+// TODO Not working. It needs review.
```
What exactly is not working?

rafaelsouzaf replied:
All the examples I am compiling (obviously) AND executing with NodeJS. TS compiler is good but some problems I can see only when I execute it. In this particular case, there is an exception in execution time. So I am not sure if the problem is my definitions, is the example or the hapijs framework.

I think the problem is with the example code. But I am not sure. I would like to confirm it with someone.

```AssertionError [ERR_ASSERTION]: Unknown cache countries
    at module.exports.internals.Core._cachePolicy (D:\Workspace\DefinitelyTyped-Rafael\node_modules\hapi\lib\core.js:520:14)
    at internals.Server.policy [as cache] (D:\Workspace\DefinitelyTyped-Rafael\node_modules\hapi\lib\server.js:513:27)
    at Object.<anonymous> (D:\Workspace\DefinitelyTyped-Rafael\types\hapi\v17\dist\server\server-cache-provision.js:10:20)
    at Module._compile (module.js:641:30)
    at Object.Module._extensions..js (module.js:652:10)
    at Module.load (module.js:560:32)
    at tryModuleLoad (module.js:503:12)
    at Function.Module._load (module.js:495:3)
    at Function.Module.runMain (module.js:682:10)
    at startup (bootstrap_node.js:191:16)
[nodemon] app crashed - waiting for file changes before starting...```

BorntraegerMarc replied:
Did you copy it from here https://github.com/hapijs/hapi/blob/master/API.md#-await-servercacheprovisionoptions ? Because one difference I see is that the example uses `await`, which we don't in our test, here...

BorntraegerMarc commented on types/hapi/v17/test/server/server-decorations.ts:
```diff
@@ -0,0 +1,25 @@
+server.start();
+server.decorate('toolkit', 'success', success);
+*/
+// TODO how to implement it with TS?
```
Not sure where the problem is here. Could you specify?

rafaelsouzaf replied:
My mistake. Fixed.

BorntraegerMarc commented on types/hapi/v17/test/server/server-info.ts:
```diff
@@ -0,0 +1,15 @@
+
+// check the correct port
+console.log(server.info);
+if (server.info) assert(server.info.port === 8000);
```
Do the `assert` checks actually get executed? I thought the tests were mainly to see if it compiles without errors? Not to actually make unit tests. Unit tests seem to be a bit of an overkill for me because the whole code is already unit tested in the "regular" hapijs project. So my assumption is -> if it compiles correctly to JS -> then the "normal" unit tests will have tested the code correctly anyway. That is why `assert`'s are not mandatory :)

rafaelsouzaf replied:
@BorntraegerMarc Yes, you are right. The most important is compiling. I am executing as well but the assert's is too much. 3d6b184

BorntraegerMarc commented on types/hapi/v17/definitions/server/server-options.d.ts:
```diff
@@ -0,0 +1,188 @@
+ * All options are optionals.
+ * [See docs](https://github.com/hapijs/hapi/blob/master/API.md#-server-options)
+ */
+export interface ServerOptions {
```
Maybe one test where all the options of this interface is used, would be nice?

rafaelsouzaf replied:
@BorntraegerMarc very good idea. I created this [test file](https://github.com/rafaelsouzaf/DefinitelyTyped/blob/6a92487939470e96f6f3219963e66962a97ff6ac/types/hapi/v17/test/server/server-options.ts).

BorntraegerMarc reviewed (approved):
So first of all let me say: Great work! I really appreciate the work you went through @rafaelsouzaf 🎉 Everyone is going to benefit from these changes :)

I agree with @gitawego -> I think the PR is in a state where it is good to be merged into the official DefinitelyTyped repository. From there we will see again some more reviews from early adopters and maybe new PR's will come up to address these reviews.

From my side: we are good to go to merge the PR 😄 

PS: Personally I like the style `response(value: string | object)` better :)

rafaelsouzaf commented:
> So first of all let me say: Great work! I really appreciate the work you went through @rafaelsouzaf 🎉 Everyone is going to benefit from these changes :)

Thanks @BorntraegerMarc . It was a very nice experience and I learned a lot. Thank you!

> I agree with @gitawego -> I think the PR is in a state where it is good to be merged into the official DefinitelyTyped repository.

Perfect. I will come back with the "method overriding" and prepare the PR to upstream.
👍 👍
